### PR TITLE
update pal api to v2

### PR DESCRIPTION
**Pull request by pangzi85480**
Signed-off-by: jack.wxz <wangxiaozhe@linux.alibaba.com>

### CODE REVIEW

File: rune/libcontainer/nsenter/loader.c [hunk 1]
```diff
@@ -28,13 +28,22 @@ struct pal_stdio_fds {
...
 };
 
+struct pal_create_process_args {
+	const char *path;
+	const char **argv;
+	const char **env;
+	const struct occlum_stdio_fds *stdio;
+	int *pid;
+};
+
```
Comment (jiazhang0) [inline rune/libcontainer/nsenter/loader.c:38]: Remove this empty line.

File: rune/libenclave/internal/runtime/pal/api_linux_v2.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+/*
+#include <stdlib.h>
+
+static int palCreateProcessV2(void *sym, const char *exe, const char *argv[],
+			    const char *envp[], int stdin, int stdout,
+			    int stderr, int *pid)
+{
+	typedef struct {
+                int stdin, stdout, stderr;
+        } pal_stdio_fds;
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v2.go:12]: Use `go fmt api_linux_v2.go` to kill all format issues.

**Comment by uudiin** [comment]
LGTM, thanks.

File: rune/libenclave/internal/runtime/pal/api_linux_v2.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+	typedef struct {
+		const char *path;
+		const char **argv;
+		const char **env;
+		pal_stdio_fds *fds;
+		int *pid;
+	} pal_create_process_args;
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v2.go:20]: Use the anonymous struct to define fds and create_process_args, such as:
```C
struct {
 		const char *path;
 		const char **argv;
 		const char **env;
 		pal_stdio_fds *fds;
 		int *pid;
} create_process_args = {
 		exe,
 		argv,
 		envp,
 		&fds,
 		pid,
};
```

File: rune/libenclave/internal/runtime/pal/api_linux_v2.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+{
+	typedef struct {
+		int pid;
+		int *exit_value;
+	} pal_exec_args;
+
+	pal_exec_args args = {
+		pid,
+		exit_code,
+	};
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v2.go:48]: ditto

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 5]
```diff
@@ -5,6 +5,10 @@ import (
 	"os"
 )
 
+const (
+	currentPalApiVersion = 2
+)
+
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:11]: Just move the definition of pal api version from api_linux.go to pal.go and define it to 2.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 6]
```diff
@@ -15,7 +19,7 @@ func (pal *enclaveRuntimePal) Load(palPath string) (err error) {
 func (pal *enclaveRuntimePal) getPalApiVersion() error {
 	api := &enclaveRuntimePalApiV1{}
 	ver := api.get_version()
-	if ver > palApiVersion {
+	if ver > currentPalApiVersion {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:22]: So we don't need to change here.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 7]
```diff
@@ -32,7 +36,12 @@ func (pal *enclaveRuntimePal) Attest() (err error) {
 }
 
 func (pal *enclaveRuntimePal) Exec(cmd []string, envp []string, stdio [3]*os.File) (int32, error) {
-	api := &enclaveRuntimePalApiV1{}
+	if pal.version <= 1 {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:39]: Change this to pal.version == 1. In order to handle pal.version == 0, add a check in getPalApiVersion().

File: rune/libenclave/internal/runtime/pal/api_linux_v2.go (hunk 3 above)
Comment (uudiin) [reply rune/libenclave/internal/runtime/pal/api_linux_v2.go:20]: pal_create_process_args is used to define the function pointer type, not anonymous.

**Comment by jiazhang0** [comment]
This commit is re-submitted by https://github.com/alibaba/inclavare-containers/pull/45 .

**Pull request closed without merging**